easybuild/tools/downloadtools.py

```diff
@@ -0,0 +1,187 @@
+import shutil
+import sys
+import urllib2
+
+from vsc.utils import fancylogger
+
+from easybuild.tools import build_log  # import build_log must stay, to activate use of EasyBuildLog
+from easybuild.tools.config import build_option
+from easybuild.tools.repository.svnrepo import SvnRepository
+from easybuild.tools.filetools import make_tarfile, write_file, mkdir
+
+
+_log = fancylogger.getLogger('filetools', fname=False)
+
+# list of possible protocols we should know how to download
+HTTP = 'download_http'
+# GIT = download_git
+SVN = 'download_svn'
+# BZR = download_bzr
+FILE = 'download_local_file'
+# FTP = download_ftp
+
+# map protocols to things we know how to download
+PROTOCOL_MAP = {
+    'http': HTTP,
+    'https': HTTP,
+    # 'http+git': GIT,
+    # 'git': GIT,
+    # 'bzr': BZR,
+    # 'ftp': FTP,
+    'svn': SVN,
+    'svn+http': SVN,
+    'file': FILE,
+}
+
+
+def download_http(filename, url, path, timeout):
+    """
+    Download a file over the http protocol
+    """
+    # try downloading, three times max.
+    downloaded = False
+    max_attempts = 3
+    attempt_cnt = 0
+
+    while not downloaded and attempt_cnt < max_attempts:
+        try:
+            # urllib2 does the right thing for http proxy setups, urllib does not!
+            url_fd = urllib2.urlopen(url, timeout=timeout)
+            _log.debug('response code for given url %s: %s' % (url, url_fd.getcode()))
+            write_file(path, url_fd.read())
+            _log.info("Downloaded file %s from url %s to %s" % (filename, url, path))
+            downloaded = True
+            url_fd.close()
+        except urllib2.HTTPError as err:
+            if 400 <= err.code <= 499:
+                _log.warning("url %s was not found (HTTP response code %s), not trying again" % (url, err.code))
+                break
+            else:
+                _log.warning("HTTPError occured while trying to download %s to %s: %s" % (url, path, err))
+                attempt_cnt += 1
+        except IOError as err:
+            _log.warning("IOError occurred while trying to download %s to %s: %s" % (url, path, err))
+            attempt_cnt += 1
+        except Exception, err:
+            _log.error("Unexpected error occurred when trying to download %s to %s: %s" % (url, path, err))
+
+        if not downloaded and attempt_cnt < max_attempts:
+            _log.info("Attempt %d of downloading %s to %s failed, trying again..." % (attempt_cnt, url, path))
+
+    if downloaded:
+        _log.info("Successful download of file %s from url %s to path %s" % (filename, url, path))
+        return path
+    else:
+        _log.warning("Download of %s to %s failed, done trying" % (url, path))
+        return None
+
+
+def download_svn(revision, url, path, timeout):
+    """
+    Download a svn repository and create a tarball out of it in the given path
+    timeout is ignored here
+    """
+    # transform url into something SvnRepository understands
+    if url.startswith('svn+'):
```

Comment:
don't hardcode the `SVN+` bit, use a constant (and then use that constant in other places where you're using `svn+`

no magic strings, please